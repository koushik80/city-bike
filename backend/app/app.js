// external imports
const express = require("express");
const app = express();
const dotenv = require("dotenv");
const cors = require("cors");

// dot environment config
if (process.env.NODE_ENV !== "PRODUCTION") {
    dotenv.config({ path: 'backend/config/config.env' });
}

// parsing json
app.use(express.json());

// used cors to allow cross-origin access with frontend
app.use(cors());

// imported routers (internal imports)
const journey = require("../routes/journey.routes");
const station = require("../routes/station.routes");
const { errorMiddleware } = require("../middleware/errorMiddleware");

// middleware
app.use("/api/journey", journey);
app.use("/api/station", station);


app.get('/', (req, res) => {
    res.status(200).json({
        message: "This is JOURNEY REST API"
    });
});

// ERROR MIDDLEWARE
app.use(errorMiddleware)

module.exports = app;