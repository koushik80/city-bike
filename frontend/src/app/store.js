// https://redux-toolkit.js.org/rtk-query/overview

import { configureStore } from '@reduxjs/toolkit'
import apiSlice from '../features/api/apiSlice'
import filterReducer from '../features/filter/filterSlice'

export const store = configureStore({
  reducer: {
    [apiSlice.reducerPath]: apiSlice.reducer,
    filter: filterReducer
  },
  devTools: process.env.NODE_ENV !== 'production',
  middleware: (getDefaultMiddlewares) => getDefaultMiddlewares().concat(apiSlice.middleware)
})