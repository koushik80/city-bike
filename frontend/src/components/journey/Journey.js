import React from 'react';
import './Journey.scss';

const Journey = () => {
  return (
    <div>

    </div>
  );
};

export default Journey;