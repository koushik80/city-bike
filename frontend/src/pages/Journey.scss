.journey-page {
    &--search-section {
        width: 100%;
        display: flex;
        align-items: center;
        justify-content: center;
        background-color: #5670ff;
        padding: 1rem 0rem;
        margin: 2rem 0rem;
        border-radius: 3px;

        &--div {
          display: flex;

          input {
            width: 300px;
            padding: 8px 16px;
            outline: none;
            font-size: 18px;
            border-radius: 3px 0px 0px 3px;
          }

          button{
            padding: 8px 16px;
            border-radius: 0px 3px 3px 0px;
          }
        }
      }
}