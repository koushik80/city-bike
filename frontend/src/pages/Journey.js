import { Box, Card, Container, Pagination, Stack, Typography } from '@mui/material';
import { DataGrid } from '@mui/x-data-grid';
import React, { Fragment, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { journeyPagination, journeySearch } from '../features/filter/filterSlice';
import { useGetJourneyQuery } from '../features/journey/journeyApi';
import { meterToKm, secondsToHours } from '../utils/utils';
import './Journey.scss';

const Journey = () => {

  const dispatch = useDispatch();
  const debounceHandler = (fn, delay) => {
    let timeoutId;
    return (arg) => {
      if (timeoutId) {
        clearTimeout(timeoutId)
      }
      timeoutId = setTimeout(() => {
        fn(arg)
      }, delay);
    }
  };

  const doSearch = (text) => {
    dispatch(journeySearch(text))
  };

  const handleSearch = debounceHandler(doSearch, 500);

  const [page, setPage] = useState(1);

  const handleChange = (event, value) => {
    setPage(value);
    dispatch(journeyPagination(value));
  };

  const query = useSelector(state => state.filter);

  let queryStr = `?page=${query?.journeyPage}`;

  if (query?.journeySearch !== undefined) {
    queryStr = `?keyword=${query?.journeySearch}&page=${query?.journeyPage}`;
  } else if (query?.journeySearch && query?.journeyPage) {
    queryStr = `?keyword=${query?.journeySearch}&page=${query?.journeyPage}`;
  } else {
    queryStr = `?page=${query?.journeyPage}`;
  }

  const { data, isLoading } = useGetJourneyQuery(queryStr);

  const columns = [
    {
      field: "_id", headerName: "ID", winWidth: 100, flex: 0.2,
      renderCell: (params) => {
        return (
          <Fragment>
            <Typography>
               {(params?.row?._id).slice(15)}
            </Typography>
          </Fragment>
        )
      }
      },
    {
      field: "departure_station_name", headerName: "Departure Name", winWidth: 120, flex: 0.3
    },
    {
      field: "return_station_name", headerName: "Return Name", winWidth: 120, flex: 0.3
    },
    {
      field: "covered_distance", headerName: "Distance (Km)", winWidth: 120, flex: 0.2,
      renderCell: (params) => {
        return (
          <Fragment>
            <Typography>
               {meterToKm(params?.row?.covered_distance).toFixed(1)} Km
            </Typography>
          </Fragment>
        )
      }
    },
    {
      field: "duration", headerName: "Duration (s)", winWidth: 100, flex: 0.2,
      renderCell: (params) => {
        return (
          <Fragment>
            <Typography>
               {secondsToHours(params?.row?.duration)}
            </Typography>
          </Fragment>
        )
      }
    },
    {
      field: "departure", headerName: "Departure Time", winWidth: 100, flex: 0.2,
      renderCell: (params) => {
        return (
          <Fragment>
            <Typography>
               {(new Date(params?.row?.return)).toLocaleDateString()}
            </Typography>
          </Fragment>
        )
      }
    },
    {
      field: "return", headerName: "Return Time", winWidth: 100, flex: 0.2,
      renderCell: (params) => {
        return (
          <Fragment>
            <Typography>
               {(new Date(params?.row?.return)).toLocaleDateString()}
            </Typography>
          </Fragment>
        )
      }
    },
  ]

  return (
    <Container>
      <div className='journey-page--search-section'>
          <form action="" onSubmit={(e) => e.preventDefault()}>
             <div className='journey-page--search-section--div'>
                <input
                    type="search" placeholder='search by departure station name'
                    onChange={({ target }) => handleSearch(target.value)}
                    onBlur={({ target }) => handleSearch(target.value)}
                />
                <button type="submit"><i className="fa-solid fa-magnifying-glass"></i></button>
             </div>
          </form>
        </div>
      <Stack direction="row" alignItems="center" justifyContent="center" mb={1}>
        <Typography variant="h5" gutterBottom>
            Journey list view
        </Typography>
      </Stack>
      <Card>
        <Box
          sx={{
            minHeight: 400,
            width: 1
          }}
        >
          <DataGrid
              loading={!data?.journey?.length || isLoading}
              getRowId={(row) => row?._id}
              columns={columns}
              pageSize={100}
              autoHeight
              rows={data?.journey || []}
          />
        </Box>
      </Card>
      <div className="station-page--pagination">
        {
          data?.journey?.length === 100 ? <Pagination count={Math.round(data?.totalJourney / 100) || 0} variant="outlined" color="primary" page={page} onChange={handleChange} /> : ""
        }
      </div>
    </Container>
  );
};

export default Journey;