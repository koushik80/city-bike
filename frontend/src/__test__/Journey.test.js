import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { store } from '../app/store';
import Journey from '../pages/Journey';

describe("render journey component", () => {

    it("should match journey component snapshot", () => {
        const wrapper = renderer.create(<Provider store={store}><Journey /></Provider>);

        expect(wrapper).toMatchSnapshot();
    })
});