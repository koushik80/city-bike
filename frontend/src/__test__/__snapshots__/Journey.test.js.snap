// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render journey component should match journey component snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
>
  <div
    className="journey-page--search-section"
  >
    <form
      action=""
      onSubmit={[Function]}
    >
      <div
        className="journey-page--search-section--div"
      >
        <input
          onBlur={[Function]}
          onChange={[Function]}
          placeholder="search by departure station name"
          type="search"
        />
        <button
          type="submit"
        >
          <i
            className="fa-solid fa-magnifying-glass"
          />
        </button>
      </div>
    </form>
  </div>
  <div
    className="css-1macvdz-MuiStack-root"
  >
    <h5
      className="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
    >
      Journey list view
    </h5>
  </div>
  <div
    className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-bhp9pd-MuiPaper-root-MuiCard-root"
  >
    <div
      className="MuiBox-root css-2jauji"
    >
      <div
        aria-colcount={7}
        aria-multiselectable={false}
        aria-rowcount={1}
        className="MuiDataGrid-root MuiDataGrid-autoHeight MuiDataGrid-root--densityStandard css-1e2bxag-MuiDataGrid-root"
        role="grid"
      >
        <div
          className="MuiDataGrid-main css-204u17-MuiDataGrid-main"
        >
          <div
            className="MuiDataGrid-overlay css-f7tavo-MuiDataGrid-overlay"
          >
            Cannot set properties of null (setting 'scrollLeft')
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="station-page--pagination"
  />
</div>
`;
